server/push/test/test_push.js

```diff
@@ -0,0 +1,133 @@
+    'rubfijq60eVuDPmfyh-fpYpOqc=',
+    'auth': 'cQXITd9TDD16BxL8qYHBdQ==',
+  },
+};
+
+const TEST_PUSH_TOKENS_PC_FIREFOX = {
+  endpoint: 'https://updates.push.services.mozilla.com/wpush/v2/gAAAAABZrVyI' +
+  'bKT4XCMsAL0qc9MOiQ_VieAYSOuy6SNUsF1mLqnEiSsmtmqiBu0COHwkcVrTk3-Sh6n00YBzR' +
+  'zY_0U4DUOJMfp52KXQPt75jpJUx3tFgOIfLJLaJP3OsSKjSHvxlc55wbDTJu_wd_1rGAus29J' +
+  '6RzdsQ05xX9n_ro4ITnQ4FLSw',
+  keys: {
+    'p256dh': 'BOsO8MAUwyfjRuXdViVv2Df58W4C-nPWIBFRnCTo-HhGUV2xjcTbwLWlSpsM3' +
+    'as2Bx0lzlE5kvPd-sjXjzy0qIo',
+    'auth': '0j5jZtzgM6L3OVQJOwSOzg',
+  },
+};
+
+const TEST_PAYLOAD = {
+  title: 'Absolute Test Payload',
+  body: 'Push Test body :)',
+};
+
+describe('[POSITIVE] Push Send', () => {
+  it('Send push with fixed push tokens - pc / chrome', () => {
+    return pushSender.sendPush(TEST_PUSH_TOKENS_PC_CHROME.endpoint,
+      TEST_PUSH_TOKENS_PC_CHROME.keys,
+      JSON.stringify(TEST_PAYLOAD))
+      .then((result) => {
+        assert.equal(result.statusCode, 201);
+      }).catch((error) => {
+        assert.fail('Failed to sendPush! ' + error);
+      });
+  });
+
+  it('Send push with fixed push tokens - pc / firefox', () => {
+    return pushSender.sendPush(TEST_PUSH_TOKENS_PC_FIREFOX.endpoint,
+      TEST_PUSH_TOKENS_PC_FIREFOX.keys, JSON.stringify(TEST_PAYLOAD))
+      .then((result) => {
+        assert.equal(result.statusCode, 201);
+      }).catch((error) => {
+        assert.fail('Failed to sendPush! ' + error);
+      });
+  });
+  // TODO(jaychl): mobile chrome / other browser token
+  // should be added for testing device
+  // TODO(jaychl): need to add user id and tokens to db then
+  // sending by userid after starting server
+});
+
+describe('[NEGATIVE] Push Send', () => {
+  const INVALID_P256_KEYS = {
+    'p256dh': 'bCgobso9T0YT3u449ro9GE6RXuobJet6EhAcEpj5m23ms38vrEIBFKxeY_yl3' +
+    '3ULYF_btMdUy0c9mhipqAMuXL0=',
+    'auth': 'BZa9OSqugZMNdSADtoP-9A==',
+  };
+  const INVALID_AUTH_KEYS = {
+    'p256dh': 'BGCibcbrMIdfz34rG4IAIfuhmvkLBEAE7l_S4zjVRjWkN4_bxo5oXO1a_0zYP' +
+    'rubfijq60eVuDPmfyh-fpYpOqc=',
+    'auth': 'CQXITd9TDD16BxL8qYHBdQ==',
+  };
+  const INVALID_SHORT_AUTH_KEYS = {
+    'p256dh': 'BCgobso9T0YT3u449ro9GE6RXuobJet6EhAcEpj5m23ms38vrEIBFKxeY_yl3' +
+    '3ULYF_btMdUy0c9mhipqAMuXL0=',
+    'auth': 'bZa9OSqugZMNdSADtoP-==',
+  };
+
+  it('Send push with invalid endpoint', () => {
+    return pushSender.sendPush('https://android.googleapis.com/gcm/send/c6cb_:',
+     TEST_PUSH_TOKENS_PC_CHROME.keys, JSON.stringify(TEST_PAYLOAD))
+      .then((result) => {
+        assert.notEqual(result.statusCode, 201);
+      }).catch((error) => {
+        assert.equal(error.statusCode, 400);
+      });
+  });
+
+  it('Send push with invalid P256dh key', () => {
+    return pushSender.sendPush(TEST_PUSH_TOKENS_PC_CHROME.endpoint,
+      INVALID_P256_KEYS, JSON.stringify(TEST_PAYLOAD))
+      .then((result) => {
+        assert.notEqual(result.statusCode, 201);
```

Comment:
I'm not sure but this shouldn't happen, right?
If so, shouldn't we assert error here?

Reply:
yes, it shouldn't happen.  So it should not get 201(normal case) and normally it is ok catching error.
(browsers can have different behavior. some browser does not catching error)